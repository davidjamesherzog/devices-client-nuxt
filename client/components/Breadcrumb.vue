<template>
  <nav class="breadcrumb" aria-label="breadcrumbs" data-breadcrumbs>
    <ul>
      <li data-breadcrumb="home" class="capitalize">
        <nuxt-link to="/">home</nuxt-link>
      </li>
      <li
        v-for="(path, key) of paths"
        :key="key"
        class="capitalize"
        :class="{ 'is-active': key === paths.length - 1 }"
      >
        <nuxt-link
          :data-breadcrumb="path"
          :to="constructUrl(path)"
          aria-current="page"
        >
          {{ path }}
        </nuxt-link>
      </li>
    </ul>
  </nav>
</template>

<script lang="ts">
import { Component, Watch, Vue } from 'nuxt-property-decorator';

@Component({
  name: 'Breadcrumb'
})
export default class Breadcrumb extends Vue {
  private paths: String[] = [];

  public created(): void {
    this.updateBreadcrumbs();
  }

  @Watch('$route')
  public onRouteChange() {
    this.updateBreadcrumbs();
  }

  public updateBreadcrumbs(): void {
    if (!this.$route || !this.$route.path || this.$route.path === '/') {
      this.paths = [];
      return;
    }
    this.paths = this.$route.path.split('/');
    if (!this.paths[0]) {
      this.paths.shift();
    }
  }

  public constructUrl(path: string): string {
    const index = this.paths.findIndex((value) => value === path);
    const paths = this.paths.slice(0, index + 1);
    return `/${paths.join('/')}`;
  }
}
</script>

<style type="scss" scoped>
.breadcrumb {
  margin-left: 1.5em;
}

.capitalize {
  text-transform: capitalize;
}
</style>
