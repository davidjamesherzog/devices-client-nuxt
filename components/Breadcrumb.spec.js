import { createLocalVue, shallowMount, RouterLinkStub } from '@vue/test-utils';
import Buefy from 'buefy';
import Breadcrumb from '~/components/Breadcrumb.vue';

describe('Breadcrumb', () => {
  describe('created', () => {
    test('should call updateBreadcrumbs', () => {
      const breadcrumb = new Breadcrumb();
      const mockContext = {
        updateBreadcrumbs: jest.fn()
      };
      breadcrumb.$options.created[0].bind(mockContext)();
      expect(mockContext.updateBreadcrumbs).toHaveBeenCalled();
    });
  });

  describe('onRouteChange', () => {
    test('should call updateBreadcrumbs', () => {
      const breadcrumb = new Breadcrumb();
      breadcrumb.updateBreadcrumbs = jest.fn();
      breadcrumb.onRouteChange();
      expect(breadcrumb.updateBreadcrumbs).toHaveBeenCalled();
    });
  });

  describe('updateBreadcrumbs', () => {
    test('should set empty paths when no $route', () => {
      const breadcrumb = new Breadcrumb();
      breadcrumb.updateBreadcrumbs();
      expect(breadcrumb.paths).toEqual([]);
    });

    test('should set empty paths when no $route.path', () => {
      const breadcrumb = new Breadcrumb();
      breadcrumb.$route = {};
      breadcrumb.updateBreadcrumbs();
      expect(breadcrumb.paths).toEqual([]);
    });

    test('should set empty paths when $route.path = /', () => {
      const breadcrumb = new Breadcrumb();
      breadcrumb.$route = {
        path: '/'
      };
      breadcrumb.updateBreadcrumbs();
      expect(breadcrumb.paths).toEqual([]);
    });

    test('should set paths', () => {
      const breadcrumb = new Breadcrumb();
      breadcrumb.$route = {
        path: 'devices/1'
      };
      breadcrumb.updateBreadcrumbs();
      expect(breadcrumb.paths).toEqual(['devices', '1']);
    });

    test('should set paths with preceding /', () => {
      const breadcrumb = new Breadcrumb();
      breadcrumb.$route = {
        path: '/devices/1'
      };
      breadcrumb.updateBreadcrumbs();
      expect(breadcrumb.paths).toEqual(['devices', '1']);
    });
  });

  describe('constructUrl', () => {
    test('should construct URL', () => {
      const breadcrumb = new Breadcrumb();
      breadcrumb.paths = ['devices', '1'];
      const url = breadcrumb.constructUrl('1');
      expect(url).toEqual('/devices/1');
    });
  });

  describe('template', () => {
    let localVue;

    beforeEach(() => {
      localVue = createLocalVue();
      localVue.use(Buefy);
    });

    test('is a Vue instance', () => {
      const wrapper = shallowMount(Breadcrumb, {
        localVue,
        stubs: {
          NuxtLink: RouterLinkStub
        }
      });
      expect(wrapper.vm).toBeTruthy();
      expect(wrapper).toMatchSnapshot();
    });

    test('should render home breadcrumb', () => {
      const $route = {
        path: ''
      };
      const wrapper = shallowMount(Breadcrumb, {
        localVue,
        mocks: {
          $route
        },
        stubs: {
          NuxtLink: RouterLinkStub
        }
      });
      const breadcrumb = wrapper.find('[data-breadcrumb="home"]');
      expect(breadcrumb.exists()).toBeTruthy();
      expect(breadcrumb.text()).toEqual('home');
      expect(wrapper).toMatchSnapshot();
    });

    test('should render networks breadcrumb', () => {
      const $route = {
        path: 'networks'
      };
      const wrapper = shallowMount(Breadcrumb, {
        localVue,
        mocks: {
          $route
        },
        stubs: {
          NuxtLink: RouterLinkStub
        }
      });
      const breadcrumb = wrapper.find('[data-breadcrumb="networks"]');
      expect(breadcrumb.exists()).toBeTruthy();
      expect(breadcrumb.text()).toEqual('networks');
      expect(wrapper).toMatchSnapshot();
    });
  });
});
