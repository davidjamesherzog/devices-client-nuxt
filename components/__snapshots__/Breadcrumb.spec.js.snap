// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumb template is a Vue instance 1`] = `
<nav aria-label="breadcrumbs" data-breadcrumbs="" class="breadcrumb">
  <ul>
    <li data-breadcrumb="home" class="capitalize"><a>home</a></li>
  </ul>
</nav>
`;

exports[`Breadcrumb template should render home breadcrumb 1`] = `
<nav aria-label="breadcrumbs" data-breadcrumbs="" class="breadcrumb">
  <ul>
    <li data-breadcrumb="home" class="capitalize"><a>home</a></li>
  </ul>
</nav>
`;

exports[`Breadcrumb template should render networks breadcrumb 1`] = `
<nav aria-label="breadcrumbs" data-breadcrumbs="" class="breadcrumb">
  <ul>
    <li data-breadcrumb="home" class="capitalize"><a>home</a></li>
    <li class="capitalize is-active"><a data-breadcrumb="networks" aria-current="page">
        networks
      </a></li>
  </ul>
</nav>
`;
